Added `IDefVar` to list of indexable definitions for defs db

callendorph opened the pull request:
Previously we just had `IDef` which meant that `val` statements would export but `var` statements would not.

I've added `IDefVar` to `Indexable` and I've also created a new `DefinitionAnnotation` for `SrcDefVariable` so that we can differentiate between `val` (immutable) and `var` (mutable).

I've tested this locally on OS-X and was able to get both `val` and `var` definitions as well as the `mutable?` flag giving the right value.

CuppoJava commented on compiler/stz-reg-alloc.stanza:
```diff
@@ -204,6 +204,8 @@ defpackage stz/reg-alloc :
 ;===================== Driver ===============================
 ;============================================================
 
+val SMALL-SMAP-SIZE = 16 ;; SIZE OF SMAP WHICH CAN BE CALLED WITHOUT CHECKING STACK-SIZE
```
Create a 
```
;====================================
;========== PARAMETERS ===============
;====================================
```

section at the top of the file and put this definition there. 

Please follow code comment conventions. Comments start with a single semi-colon are capitalized and then lower-case.

jackbackrack replied:
done.

CuppoJava commented on compiler/stz-reg-alloc.stanza:
```diff
@@ -2681,6 +2683,31 @@ defn assemble (emitter:CodeEmitter, stackmap:StackMap, stubs:AsmStubs) :
         val amb-branch = label-table[amb(e)]
         E $ asm-MethodDispatch(multi(e), num-header-args(e), no-branch, amb-branch)        
       (e) : fatal("Not yet implemented: %_" % [e])
+    
```
This logic is too fragile. Let's introduce a new instruction called:

```
SectionMarker :
  name: Symbol
```

And have the preamble be surrounded by a pair of `SectionMarker` instructions. Then strip away everything in that section.

jackbackrack replied:
done.

CuppoJava commented on compiler/stz-reg-alloc.stanza:
```diff
@@ -2436,7 +2438,7 @@ defn assemble (emitter:CodeEmitter, stackmap:StackMap, stubs:AsmStubs) :
             val return-lbl = unique-id(stubs)
             E $ SetL(R0, LM(return-lbl))
-            E $ SetL(R1, INT(size(stackmap) + 8))
+            E $ SetL(R1, INT(size(stackmap) + (8 + SMALL-SMAP-SIZE)))
```
This might be the right logic, but before I can merge it I need to refresh my head about the details of the stack layout. It's tricky, and if we get it wrong, it manifests as random crashes every few weeks, so it's important we get it right on the first try.

To do that, I need to know what is the stack layout, what is enforced by this instruction, why that `+8` is there, etc.

jackbackrack replied:
the +8 is yours.  i just added SMALL-SMAP-SIZE.

CuppoJava commented on compiler/stz-compiler-main.stanza:
```diff
@@ -179,6 +179,7 @@ public defn compile (proj-manager:ProjManager,
     switch(situation) :
       `optimized-asm :
         val packages = Vector<VMPackage|StdPkg>()
+        val opt-timer = MillisecondTimer("Opt")
```
Please cleanup before merge.

callendorph committed:
Added `IDefVar` to list of indexable definitions for defs db

Previously we just had `IDef` which meant that `val` statements would
export but `var` statements would not.

I've added `IDefVar` to `Indexable` and I've also created a new
`DefinitionAnnotation` for `SrcDefVariable` so that we can differentiate
between `val` (immutable) and `var` (mutable).
